website/docs/guides/container_service_extension_4_0.html.markdown

```diff
@@ -0,0 +1,411 @@
+  - OVAs are quite big, you could tune `upload_piece_size` to speed up the upload process.
+  - If upload fails, or you need to re-upload it, you can do a `terraform apply -replace=vcd_catalog_vapp_template.cse_ova`.
+  - Verify that there's not a huge latency between your VCD and the place where Terraform configuration is run.
+
+- Cluster creation is failing:
+  - Please visit the [CSE documentation][cse_docs] to learn how to monitor the logs and troubleshoot possible problems.
+
+## Update CSE Server
+
+### Update Configuration
+
+To make changes to the existing server configuration, you should be able to locate the [`vcd_rde`][rde] resource named `vcdkeconfig_instance`
+in the [proposed configuration][step2] that was created during the installation process. To update its configuration, you can
+**change the variable values that are referenced**. For this, you can review the **"CSE Server"** section in the Installation process to
+see how this can be done.
+
+After variables are changed, the CSE Server VM needs to be rebooted. You can trigger a reboot changing the CSE Server configuration
+as follows:
+
+```hcl
+resource "vcd_vapp_vm" "cse_server_vm" {
+  # ...
+  power_on = false # Trigger a power off
+  # ...
+}
+```
+
+Then change to power on again:
+
+```hcl
+resource "vcd_vapp_vm" "cse_server_vm" {
+  # ...
+  power_on = true # Trigger a power on
+  # ...
+}
+```
+
+This must be done as a 2-step operation.
+
+### Patch version upgrade
+
+To upgrade the CSE Server appliance, first you need to upload a new CSE Server OVA to the CSE catalog and then replace
+the reference to the [vApp Template][catalog_vapp_template] in the CSE Server VM.
+
+In the [proposed configuration][step2], you can find the `cse_ova` [vApp Template][catalog_vapp_template] and the 
+`cse_server_vm` [VM][vm] that were applied during the installation process.
+Then you can create a new `vcd_catalog_vapp_template` and modify `cse_server_vm` to reference it:
+
+```hcl
+# Uploads a new CSE Server OVA. In the example below, we upload version 4.0.2
+resource "vcd_catalog_vapp_template" "new_cse_ova" {
+  org        = vcd_org.solutions_organization.name # References the Solutions Organization
+  catalog_id = vcd_catalog.cse_catalog.id          # References the CSE Catalog
+
+  name        = "VMware_Cloud_Director_Container_Service_Extension-4.0.2"
+  description = "VMware_Cloud_Director_Container_Service_Extension-4.0.2"
+  ova_path    = "/home/bob/cse/VMware_Cloud_Director_Container_Service_Extension-4.0.2.ova"
+}
+
+# ...
+
+# Update the vApp Template reference to update the CSE Server
+resource "vcd_vapp_vm" "cse_server_vm" {
+  # ...
+  vapp_template_id = vcd_catalog_vapp_template.new_cse_ova.id # Change to the new OVA version
+}
+```
+
+## Uninstall CSE
+
+~> Before uninstalling CSE, make sure you mark all clusters for deletion in VCD UI.
+
+Once all clusters are removed in the background by CSE Server, you may destroy the remaining infrastructure.
```

Comment:
It's awesome you have covered not only setup, but also upgrade and destroy. Tiny suggestion to be clear destroy is done with Terraform too:

```suggestion
Once all clusters are removed in the background by CSE Server, you may destroy the remaining infrastructure with Terraform command.
```

Reply:
Changed, thanks!